### PR TITLE
[DeepSeek-V4] Implement model integration, decoders, and configuration stack

### PR DESCRIPTION
**Pull request by parambole**
Description
===========

Implement full model architecture, decoder integration layers, and execution configurations required for DeepSeek-V4 integration into MaxText:

-   `deepseek_v4.py`: Model architecture definition supporting cyclical layer stacking, multi-head hyper-connections (mHC), and engram memory management.
-   `decoders.py` & `nnx_decoders.py`: Integration of `DeepSeekV4DecoderLayer`, supporting `get_attention_type` routing and scanned vs unrolled compilation parity.
-   `mhc.py` & `engram.py`: Execution layers for additive hyper-connections and engram state projection.
-   Configuration: Register model configs (`deepseek_v4-flash.yml`, `deepseek_v4-tiny.yml`) and hyperparameter definitions in `base.yml` and `types.py`.
-   Unit test suite (`tests/unit/deepseek_v4_vs_reference_test.py`) validating end-to-end decoder block parity against PyTorch reference implementations at `atol=1e-5, rtol=1e-5`.

Tests
=====

Tested on CPU

```
pytest  tests/unit/deepseek_v4_vs_reference_test.py

================= 12 passed, 318 warnings in 63.46s (0:01:03) ==================
tests/unit/deepseek_v4_vs_reference_test.py ............                 [100%]
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/models/deepseek_v4-flash.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+compress_rope_theta: 160000.0
+index_head_dim: 128
+index_n_heads: 64
+index_topk: 512
+o_groups: 8
+o_lora_rank: 1024
+sliding_window: 128
+num_hash_layers: 3
+mlp_activations_limit: 10.0
+compress_ratios: [0, 0, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4]
```
Comment (entrpn) [inline src/maxtext/configs/models/deepseek_v4-flash.yml:62]: why is the last compress ratio being excluded? HF compress ratios:

```
"compress_ratios": [0, 0, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 128, 4, 0]
}
```

File: src/maxtext/configs/models/deepseek_v4-flash.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+# Compressed Sparse Attention
+q_lora_rank: 1024
+kv_lora_rank: 512
+qk_nope_head_dim: 128
+qk_rope_head_dim: 64
+v_head_dim: 128
+mscale: 1.0
+
+# RoPE
+rope_type: "default"
```
Comment (entrpn) [inline src/maxtext/configs/models/deepseek_v4-flash.yml:73]: should this be yarn?
